tools/clang/lib/SPIRV/RemoveBufferBlockVisitor.cpp

```diff
@@ -0,0 +1,81 @@
+//===-- RemoveBufferBlockVisitor.cpp - RemoveBufferBlock Visitor -*- C++ -*-==//
+//
+//                     The LLVM Compiler Infrastructure
+//
+// This file is distributed under the University of Illinois Open Source
+// License. See LICENSE.TXT for details.
+//
+//===----------------------------------------------------------------------===//
+
+#include "RemoveBufferBlockVisitor.h"
+#include "clang/SPIRV/SpirvContext.h"
+
+namespace {
+
+bool isBufferBlockDecorationDeprecated(
+    const clang::spirv::SpirvCodeGenOptions &opts) {
+  return opts.targetEnv.compare("vulkan1.2") >= 0;
+}
+
+} // end anonymous namespace
+
+namespace clang {
+namespace spirv {
+
+bool RemoveBufferBlockVisitor::visit(SpirvModule *mod, Phase phase) {
+  // If the target environment is Vulkan 1.2 or later, BufferBlock decoration is
+  // deprecated and should be removed from the module.
+  // Otherwise, no action is needed by this IMR visitor.
+  if (phase == Visitor::Phase::Init)
+    if (!isBufferBlockDecorationDeprecated(spvOptions))
+      return false;
+
+  return true;
+}
+
+bool RemoveBufferBlockVisitor::visitInstruction(SpirvInstruction *inst) {
+  if (!inst->getResultType())
+    return true;
+
```

Comment:
So .. are `SpirvAccessChain` and `SpirvPointerType` only cases that require update?

Reply:
I believe so. OpVariable (and any other operation with a pointer result type) should be fixed by this. For `OpImageTexelPointer`, the storage class of the pointer is `Image`, so no change is needed for it.
And this code ensures that OpAccessChain generates correct storage class for its `base` (even if its base is another OpAccessChain).

Reply:
for example: 
https://github.com/microsoft/DirectXShaderCompiler/blob/3e5ab5c56b92990f9b45acc0d81df49bd7285707/tools/clang/test/CodeGenSPIRV/vk.1p2.block-decoration.hlsl#L31-L32

Reply:
Got it.